### PR TITLE
Packed storage for arrays.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: libsolidity/ExpressionCompiler.cpp [hunk 1]
```diff
@@ -811,36 +812,50 @@ bool ExpressionCompiler::visit(IndexAccess const& _indexAccess)
...
-				m_context << u256(0); // @todo
+				m_context << eth::Instruction::SWAP1;
+				if (arrayType.getBaseType()->getStorageBytes() <= 16)
+				{
+					// stack: <data_ref> <index>
+					// goal:
+					// <ref> <byte_number> = <base_ref + index / itemsPerSlot> <(index % itemsPerSlot) * byteSize>
+					unsigned byteSize = arrayType.getBaseType()->getStorageBytes();
+					unsigned itemsPerSlot = 32 / byteSize;
+					m_context << u256(itemsPerSlot) << eth::Instruction::SWAP2;
```
Comment (LianaHus) [inline libsolidity/ExpressionCompiler.cpp]: byteSize is always > 0?

Comment (chriseth) [inline libsolidity/ExpressionCompiler.cpp]: Yes, such arrays are absolutely nonsensical and forbidden at the type level.
An assert would not hurt, though.

File: libsolidity/ArrayUtils.cpp [hunk 2]
```diff
@@ -349,3 +440,38 @@ void ArrayUtils::retrieveLength(ArrayType const& _arrayType) const
 	}
 }
 
+void ArrayUtils::incrementByteOffset(unsigned byteSize, unsigned byteOffsetPosition, unsigned storageOffsetPosition) const
```
Comment (LefterisJP) [inline libsolidity/ArrayUtils.cpp]: Prepend the arguments with `_`

File: libsolidity/ArrayUtils.h [hunk 3]
```diff
@@ -72,6 +72,12 @@ class ArrayUtils
 	void retrieveLength(ArrayType const& _arrayType) const;
 
 private:
+	/// Adds the given number of bytes to a storage byte offset counter and also increments
+	/// the storage offset if adding this number again would increase the counter over 32.
+	/// @param byteOffsetPosition the stack offset of the storage byte offset
+	/// @param storageOffsetPosition the stack offset of the storage slot offset
+	void incrementByteOffset(unsigned byteSize, unsigned byteOffsetPosition, unsigned storageOffsetPosition) const;
```
Comment (LefterisJP) [inline libsolidity/ArrayUtils.h]: prepend `_` here too

**Comment by chriseth** [comment]
Merging manually since test infrastructure is broken. Local runs of Solidity tests were green.

**Pull request merged**